src/main/java/com/nvidia/spark/rapids/jni/SparkResourceAdaptor.java

```diff
@@ -116,8 +133,13 @@ public void endRetryBlock(long threadId) {
     endRetryBlock(getHandle(), threadId);
   }
 
-  public void checkAndBreakDeadlocks() {
-    checkAndBreakDeadlocks(getHandle());
+  public synchronized void checkAndBreakDeadlocks() {
```

Comment:
Is there a possible performance penalty for making this synchronized?

Reply:
Yes there is, but it should be very small. The C++ code has locking internally. The java code wrapping this call also has locking. So the only path where there was not locking was the one that we are trying to fix. The warchdog thread.